pygrep/__init__.py

```diff
@@ -2,78 +2,129 @@
-        mods = tuple(node.module.split('.'))
-        for left, right in zip(mods, self.ident):
-            if left != right:
-                return
-        subident = self.ident[len(mods):]
+        subident = self.ident[len(mods) :]
         for alias in node.names:
-            for left, right in zip(alias.name.split('.'), subident):
-                if left != right:
-                    continue
+            parts = tuple(alias.name.split("."))
+            if parts[: len(subident)] != subident:
+                continue
             lname = alias.asname is not None and alias.asname or alias.name
-            self.importMap[lname] = mods + (alias.name, )
+            self.importMap[lname] = mods + (alias.name,)
 
     def context_visit(self, node):
         self.context.append((node.name, node.lineno))
-        self.generic_visit(node)
+        yield from self.generic_visit(node)
         self.context.pop()
+
     visit_FunctionDef = visit_ClassDef = context_visit
 
     def visit_Name(self, node):
-        ident = tuple(self.attrs) + (node.id,)
-        self.attrs.clear()
-        if ident[0] in self.importMap:
-            ident = self.importMap[ident[0]] + ident[1:]
-        if len(ident) < len(self.ident):
+        if node.id not in self.importMap:
+            return
+        ident = (*self.importMap[node.id], *self.attrs)
+        if ident[: len(self.ident)] != self.ident:
             return
-        for left, right in zip(ident, self.ident):
-            if left != right:
-                return
-        self.callback(self.path, self.context, node, ident)
+        yield (self.context, node, ident)
 
     def visit_Attribute(self, node):
         self.attrs.appendleft(node.attr)
-        self.generic_visit(node)
+        yield from self.generic_visit(node)
+        self.attrs.popleft()
 
 
-class handleIdent:
-    def __init__(self, ident, files_or_dirs, callback):
+class FindIdentifier:
+    def __init__(self, ident, files_or_dirs):
         self.ident = ident
-        self.callback = callback
-        for fd in files_or_dirs:
+        self.files_or_dirs = files_or_dirs
+
+    def __call__(self) -> Generator[tuple, None]:
+        for fd in self.files_or_dirs:
             if os.path.isdir(fd):
                 for dirpath, dirnames, filenames in walk(fd):
                     for fn in filenames:
-                        if fn.endswith('.py'):
-                            self.handleFile(os.path.join(dirpath, fn))
+                        if fn.endswith(".py"):
+                            yield from self.handleFile(os.path.join(dirpath, fn))
             else:
-                self.handleFile(fd)
+                yield from self.handleFile(fd)
 
     def handleFile(self, path):
         m = ast.parse(open(path).read(), path)
-        iv = IdentVisitor(path, self.ident, self.callback)
-        iv.visit(m)
+        iv = IdentVisitor(self.ident)
+        for t in iv.visit(m):
+            yield (path, *t)
+
+
+def main():
```

Comment:
If you want to make pygrep 'more runnable', you may want a `pygrep/__main__.py`.  That's what would be run for `python -m pygrep`, for example.  It could consist of nothing more than `import pygrep; pygrep.main()`.